File: LeetCode/Hao/82RemoveDuplicatesFromSortedListII.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        
        int last = 0; 
        ListNode t = new ListNode(0);
        t.next = head;
        ListNode result = t;
        
        for (ListNode p = head; p != null; p = p.next) {
            if (p == head) {
                last = p.val;
                if (last != p.next.val) {
                    t.next = p;
                    t = t.next;
                }
            } else if (p.val != last && (p.next == null || p.next.val != p.val)) {
                t.next = p;
                t = t.next;
                last = p.val;
            } else if (p.val != last) {
                last = p.val;
            }
        }
        t.next = null;
        
        return result.next;
        
    }
}

File: LeetCode/Liang/82RemoveDuplicatesFromSortedListII.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Assume no circle.
 **/
 
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (!head) {
            return head;    
        }
        
        ListNode *preHead = new ListNode(INT_MIN);
        ListNode *tail = preHead;
        ListNode *first = head, *second = head;
        bool isDup = false;
        while (second->next) {
            second = second->next;
            if (first->val != second->val) {
                if (!isDup) {
                    tail->next = first;
                    tail = tail->next;
                }
                isDup = false;
                first = second;
            } else {
                isDup = true;
            }
        }
        
        if (!isDup) {
            tail->next = first;
            tail = tail->next;
        }
        tail->next = nullptr;
        head = preHead->next;
        delete preHead;
        return head;
    }
};
